maven-enforcer-plugin/src/main/java/org/apache/maven/plugins/enforcer/internal/EnforcerRuleManager.java

```diff
@@ -25,33 +25,52 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
-import org.apache.maven.enforcer.rule.api.EnforcerRule;
+import org.apache.maven.enforcer.rule.api.EnforcerRuleBase;
 import org.apache.maven.execution.MavenSession;
 import org.apache.maven.plugin.MojoExecution;
 import org.apache.maven.plugin.PluginParameterExpressionEvaluator;
+import org.apache.maven.plugins.enforcer.EnforcerRuleManagerException;
+import org.codehaus.plexus.PlexusContainer;
 import org.codehaus.plexus.classworlds.realm.ClassRealm;
 import org.codehaus.plexus.component.configurator.ComponentConfigurationException;
 import org.codehaus.plexus.component.configurator.ComponentConfigurator;
 import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluator;
+import org.codehaus.plexus.component.repository.exception.ComponentLookupException;
 import org.codehaus.plexus.configuration.PlexusConfiguration;
 
 /**
  * Manage enforcer rules.
+ *
+ * @author Slawomir Jaranowski
+ * @since 3.2.0
  */
 @Named
 @Singleton
 public class EnforcerRuleManager {
 
-    @Inject
-    private Provider<MavenSession> sessionProvider;
+    private final Provider<MavenSession> sessionProvider;
 
-    @Inject
-    private Provider<MojoExecution> mojoExecutionProvider;
+    private final Provider<MojoExecution> mojoExecutionProvider;
+
+    private final ComponentConfigurator componentConfigurator;
+
+    private final PlexusContainer plexusContainer;
```

Comment:
Hm, this is really legacy. New code should no longer rely on that but rather on https://www.eclipse.org/sisu/docs/api/org.eclipse.sisu.inject/reference/org/eclipse/sisu/inject/package-summary.html

Reply:
We shouldn't support custom rules with legacy plexus annotations at all.

Reply:
So which Maven component should we use for lookup only JSR 330?

Reply:
Even in new Api we use `PlexusContainer`

https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/internal/impl/DefaultLookup.java

So now I don't see another way.